### PR TITLE
feat: Keep upstream's state in local cache #585

### PR DESCRIPTION
**Pull request by astsiapanay**
Core keeps upstream's state in local cache for deployments taken from DIAl config.

We want to maintain a consistent behavior for deployments from blob storage as well.

### CODE REVIEW

File: server/src/main/java/com/epam/aidial/core/server/config/FileConfigStore.java [hunk 1]
```diff
@@ -118,7 +118,6 @@ private void load(boolean fail) {
             }
 
             this.config = config;
-            upstreamRouteProvider.onUpdate(config);
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/config/FileConfigStore.java:121]: please remove the field

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java [hunk 2]
```diff
@@ -30,90 +27,88 @@ public class UpstreamRouteProvider {
...
+    public UpstreamRoute get(Route route) {
+        String key = getKey(route);
+        return get(key, route.getUpstreams());
     }
 
-    private static void updateRoutes(Map<String, TieredBalancer> newState, Map<String, TieredBalancer> oldState, Collection<Route> routes) {
-        for (Route route : routes) {
-            String name = route.getName();
+    private UpstreamRoute get(String key, List<Upstream> upstreams) {
+        TieredBalancer balancer = balancers.merge(key, new TieredBalancer(key, upstreams), (prev, next) -> {
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java]: use compute to not create TieredBalancer all the time, it is quite expensive. Please match the given list with the current tiered balancer

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java [hunk 3]
```diff
@@ -30,90 +27,88 @@ public class UpstreamRouteProvider {
...
-        if (previousBalancer != null) {
-            log.warn("Duplicate deployment name: {}", name);
+        return prefix + ":" + deployment.getName();
+    }
+
+    private void evictExpiredBalancers() {
+        long currentTime = System.currentTimeMillis();
+        for (Map.Entry<String, TieredBalancer> entry : balancers.entrySet()) {
+            TieredBalancer balancer = entry.getValue();
+            if (currentTime - balancer.getLastAccessTime() > IDLE_PERIOD_IN_MS) {
```
Comment (artsiomkorzun) [inline server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java]: should be updated with the same memory-semantic as you set value. So map.compute and check there.

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java (hunk 2 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java (hunk 3 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/upstream/UpstreamRouteProvider.java]: fixed

File: server/src/main/java/com/epam/aidial/core/server/config/FileConfigStore.java (hunk 1 above)
Comment (astsiapanay) [reply server/src/main/java/com/epam/aidial/core/server/config/FileConfigStore.java:121]: fixed

**Review by artsiomkorzun — DISMISSED** [review]

File: config/src/main/java/com/epam/aidial/core/config/Upstream.java [hunk 4]
```diff
@@ -14,9 +15,15 @@ public class Upstream {
     public static final int ERROR_THRESHOLD = 3;
 
     private String endpoint;
+
+    @EqualsAndHashCode.Exclude
     private String key;
+
     @JsonDeserialize(using = JsonToStringDeserializer.class)
+    @EqualsAndHashCode.Exclude
     private String extraData;
```
Comment (artsiomkorzun) [inline config/src/main/java/com/epam/aidial/core/config/Upstream.java]: These two fields are used in DeploymentController. So if they are updated and the rest fields are not, you won't update a balancer and you will use the old values of theses fields forever.

**Review by artsiomkorzun — CHANGES_REQUESTED** [review]

File: config/src/main/java/com/epam/aidial/core/config/Upstream.java (hunk 4 above)
Comment (astsiapanay) [reply config/src/main/java/com/epam/aidial/core/config/Upstream.java]: Right the router may return a cached value of Upstream class. If it returned an upstream's index there wouldn't be any issue.

Comment (astsiapanay) [reply config/src/main/java/com/epam/aidial/core/config/Upstream.java]: fixed

**Review by artsiomkorzun — APPROVED** [review]

**Comment by astsiapanay** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-590.nightly-test.deltixhub.io
>E2E tests status: success

**Pull request merged**